### ISSUE TITLE
Support for opting out of jsdoc @typedef exports

### ISSUE THREAD

**Issue by jespertheend** [microsoft/TypeScript#46011]
# Suggestion

## 🔍 Search Terms

jsdoc typedef export #43207 #23692

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Currently when using `/** @typedef {Number} Num */` the type `Num` is automatically exported when the comment is in the root scope. I'd like to prevent this from happening.

## 📃 Motivating Example 💻 Use Cases

My problem:
`FileA.js` declares `/** @typedef {Number} Num */`
`FileB.js` uses this type a bunch of times, so instead of using `/** @type {import("./FileA.js").Num} */` everywhere, it declares it once at the top of the file: `/** @typedef {import("./FileA.js").Num} Num */` and then uses `Num` everywhere in the file.

The problem is shat doing it this way in all files creates a lot of exports of the same type and it's not always clear which one is the original export. This creates a rather big list with Intellisense:
![image](https://user-images.githubusercontent.com/2737650/134426452-2b90403a-496c-402b-82c1-f08a58c74c57.png)

A woraround is to wrap the type in parentheses, this causes the typedef to be limited to that scope, but this is not always feasible. I.e:
```js
{
    /** @typedef {import("./FileA.js").Num} Num */
    /**
     * @param {Num}
     */
    export default function foo(x) { // <-- error, export needs to be top level
    
    }
}
```

**Comment by andrewbranch** [comment]
Agreed this is not ideal. Not sure how best to indicate that it shouldn’t be exported. Maybe another JSDoc tag on the comment:

```ts
/**
 * @typedef {import("./FileA.js").Num} Num
 * @local
 */
```

`@local` isn’t something listed on https://jsdoc.app. `@private` or something? We kind of try not to just make up JSDoc things willy nilly, but this feels like a problem we made so the solution will probably have to be made by us too. /cc @sandersn?

**Comment by jespertheend** [comment]
I'm guessing this would no longer be an issue if #22160 is fixed.
That would allow you to import `Num` without exporting it.

**Comment by sandersn** [comment]
I think the closest tag is `@module`. 

jsdoc.app and Closure must have the problem to some degree since they have the concept of scope. What do they do? Are `@typedef` exportable somehow? I'd like to know how those two systems handle it before changing ours.

**Comment by jespertheend** [comment]
It seems JSDoc.app is having a similar issue https://github.com/jsdoc/jsdoc/issues/1537
I think both JSDoc.app and Closure are both exporting `@typedef` to file scope just like TypeScript is currently doing.

**Comment by brendankenny** [comment]
> jsdoc.app and Closure must have the problem to some degree since they have the concept of scope. What do they do? Are `@typedef` exportable somehow? I'd like to know how those two systems handle it before changing ours.

For Closure Compiler, at least, typedefs are attached to a real value-space declaration. For example:

```js
/** @typedef {(string|Element|Text|Array<Element>|Array<Text>)} */
goog.ElementContent;
```
[Types in the Closure Type System - Typedefs](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System#typedefs)

(it's been a few years since I regularly wrote Closure-able code, but I don't believe anything has changed)

As a result, export/import for typedefs is explicit and the same as bringing in any dependency (though from https://github.com/google/closure-compiler/issues/3041 it's clear this can lead to the same issues that lead to `import type` in TypeScript)

**Comment by jespertheend** [comment]
I suppose this might be a duplicate of #22160
At least that would solve the same issue.

**Comment by jespertheend** [comment]
Best workaround I've found for this so far is to assign a type to an unused variable, and then use `typeof variable` whenever you need to use the type:
```js
/**
 * @type {{
 *  foo: number,
 *  bar: string,
 * }}
 */
let myType;

/**
 * @param {typeof myType} x
 */
function takesMyType(x) {

}
```
The downside of this that it creates an unused variable, so you might have to disable your linter every time you do this.

Other than that I'm frequently importing types within the closure of a class or function, and I'm using tons and tons of inline type imports, which results in some truly horrendous looking code as can be seen in https://github.com/microsoft/TypeScript/issues/22160#issuecomment-1404399423

**Comment by kungfooman** [comment]
Problem:

![image](https://user-images.githubusercontent.com/5236548/236839369-949c014f-ed29-4704-a8ee-851a0b13a763.png)

Even though it's just this `@typedef` in each file (to make it nicer...):

```js
/**
 * @typedef {import('./utils/hub.js').PretrainedOptions} PretrainedOptions
 */
```

Solution should be as simple as implementing `@private` for `@typedef`'s to not auto-export everything blindly?

**Comment by kungfooman** [comment]
I think I found the problematic PR: first `@typedef`'s were local and now they are all exported:

https://github.com/microsoft/TypeScript/pull/23723

@sandersn Did you have this `export * from ...`-in-entry-point-index-file scenario in mind when dealing with that PR?

**Comment by ljharb** [comment]
I'm running into this now; I have a CJS file with JSDoc comments that's generating types, and the resulting types have `export =` (as they should) for the CJS value, and also `export type` for the typedef, despite not explicitly exporting it - which causes tsc to error.

How can I define a type purely in jsdoc and not export it?

**Comment by andrewbranch** [comment]
@ljharb that actually sounds like a declaration emit bug separate from this feature request. You should be getting an `export =` of a namespace containing the type, merged with your `module.exports` object, like this: https://www.typescriptlang.org/play?module=1&filetype=js#code/PQKhAIAEBcE8AcCmATRAzcBvAztATgJYB2A5gL7gBiA9teCMAFCNoCuRAxtAdUeGgAoAlFjLMAttWSsANogB0iAB7xqeaNnABefgG4gA

While being able to avoid exporting a `typedef` would work around this (and I still think it would be a good feature), you’re supposed to be able to use `module.exports` and still export types without emitting invalid declarations.

Can you show how to repro the bug you’re seeing?

**Comment by ljharb** [comment]
@andrewbranch ah, fair point. in that case, https://github.com/ljharb/set-function-name/tree/tsc, `npm run emit-types && npm run tsc` should repro the failure.

**Comment by andrewbranch** [comment]
Opened #56002

**Comment by sandersn** [comment]
> @sandersn Did you have this export * from ...-in-entry-point-index-file scenario in mind when dealing with that PR?

No, definitely not.

i like the idea of a modifier like `@local` or `@private` or `@module` on `@typedef`, although I'd like it go in a place that doesn't make `@typedef` much more complicated to parse.

It's unfortunately probably too late to change the semantics to require something like `@export` on `@typedef` for it to be exported from modules.

**Comment by kungfooman** [comment]
> It's unfortunately probably too late to change the semantics to require something like `@export` on `@typedef` for it to be exported from modules.

Yea, your PR is @export'ing them since 2018, lets see both cases:

1) default `@export`... this is against "local by default" which is the standard in ESM world.

2) default `@local`... this may break some projects which started to rely on the behaviour since 2018.

I would rather aim for (2) to keep the ESM no-exporting-unless-stated behaviour "aligned", but based on legacy, we cannot simply change it. So the only option I see is to... make it an option :sweat_smile: 

So once we have an option, we need to define how to export a `@typedef`:

```
/**
 * @typedef {object} A
 * @export
 * @typedef {object} B
 */
```

Is this exporting A or B? IMO a very specific naming is causing the least ambiguity:

```
/**
 * @typedef {object} A
 * @typedef {object} B
 * @typedef {object} C
 * @export A, C - Everyone should understand that B is not exported here?
 */
```

The default option can for my sake be (1) - keep defaulting to exporting typedefs. I've seen several projects by now being affected by this, so if the fix comes in the form of an option: problem solved.

I wonder why @DanielRosenwasser gave a thumbs-down on your 2018-typedef-export PR, maybe because of this issue? Lets just discuss a bit, decide on a solution and go for it?

**Comment by jespertheend** [comment]
What about a completely new tag? Something like
```js
/**
 * @localtype {object} A
 */
```
It would
 - seem closer to a 'default local' since no extra tag is required
 - be easier to parse, since `@typedef`s and `@localtype`s can coexist along side each other in the same comment
 - not break existing code

Not sure about the name of the tag exactly, ideally it would look similar to `@typedef` but shorter somehow, to convey the idea that this new tag is the default, whereas `@typedef` seems more like an extension of that new tag.

An other idea could be to add two completely new tags and deprecate the `@typedef` tag, but I'm not sure how feasible that is. I don't think any JSDoc tags have ever been deprecated in TypeScript, let alone one as common as `@typedef`.

**Comment by sandersn** [comment]
I like `@localtype` because it's easy to specify and easy to parse (at least, no harder than `@typedef).
But for users it's not easy to discover and mysterious if you run across it in random code.

`@export`, on the other hand, implies that it mirrors all of JS exports, which is quite complex -- see @kungfooman 's example of both a separate `@export A, C` and an export modifier `@export @typedef ...`. It is, probably, the right solution in the abstract, though, since we try to have jsdoc mirror normal TS/JS constructs as much as possible.

We're definitely not able to deprecate anything because I expect a lot of JSDoc's value is in using VS Code (or even VS!) to open ancient loose .js files and get some kind of help from jsdoc comments.

**Comment by jespertheend** [comment]
One problem with `@export` though, is that it is not really clear how to prevent _any_ type from being exported.
Specifically, if you have a JSDoc comment with only a single `@typedef`.

Something like
```js
/**
 * @typedef {number} Foo
 * @export _
 */
```
to indicate that _no_ type should be exported from that block comment could work. But honestly, that seems even more mysterious than `@localtype` at that point.

If we don't want to deprecate anything, then `@export` is pretty much out of the question.

So at this point the only options I see are:

- A new separate tag. Something like `@local @typedef`, or maybe `@private @typedef`.
- A singular tag. Something like `@localtype`.

**Comment by kungfooman** [comment]
> One problem with `@export` though, is that it is not really clear how to prevent _any_ type from being exported.

That's a simple condition:

```js
const str = `
/**
 * @typedef {number} Foo
 * @export _, 123, Foo
 */
function add(a, b) {
  return a + b;
}`;
const ast = ts.createSourceFile('repl.ts', str, ts.ScriptTarget.Latest, false /*setParentNodes*/);
const {tags} = ast.statements[0].jsDoc[0];
for (const tag of tags) {
    if (tag.tagName.text === 'export') {
        const names = tag.comment.split(',').map(_ => _.trim());
        for (const name of names) {
            const typedef = tags.find(tag => tag.kind === ts.SyntaxKind.JSDocTypedefTag && tag.name.text === name);
            if (!typedef) {
                console.log(`Trying to find @typedef called ${name}, but can't find it.`);
                continue;
            }
            console.log("Exporting", typedef.name.text);
        }
    }
}
```

Output:

![image](https://github.com/microsoft/TypeScript/assets/5236548/141737c4-6ae5-4a7c-bf68-cef2f7489b10)

> We're definitely not able to deprecate anything because I expect a lot of JSDoc's value is in using VS Code (or even VS!) to open ancient loose .js files and get some kind of help from jsdoc comments.

Yes, solved by making it an option, maybe `strictJSDoc: true`? `@localtype` sounds very strange to me anyway and it also goes against everything ESM tries to fix (keeping things local *except* they are exported) to not cause any unexpected issues... exactly the kind of issues we are running into here.



> `@export`, on the other hand, implies that it mirrors all of JS exports, which is quite complex -- see @kungfooman 's example of both a separate `@export A, C` and an export modifier `@export @typedef ...`. It is, probably, the right solution in the abstract, though, since we try to have jsdoc mirror normal TS/JS constructs as much as possible.

Since JSDoc is about types, isn't it "implies that it mirrors all of TS exports"? In the beginning we can just limit it to typedef's in the same JSDoc as a "minimal working solution" and see what other ideas developers have.

Idea for even more syntax: `@export(typedef) ...` (that would bind the `export` as much as possible to the `typedef`)

**Comment by jespertheend** [comment]
> Yes, solved by making it an option, maybe `strictJSDoc: true`?

That works, but only if all your libraries are up to date and expect that option to be set. If one of your libraries doesn't have any of its types `@export`ed and you want to use some of its types, you'll be out of luck.

I agree `@localtype` feels pretty weird, and so do `@local @typedef` or `@private @typedef`.
`@export` could work, but it will be another `// @ts-check` that you have to add to every file and I'd like to avoid that where possible.

I'm thinking of what it would be like to migrate an existing project, and adding a tag to every file just to opt out of type exporting is not really something I'd want to do. You might forget to set it in one file and accidentally export all your types without being aware of it.

**Comment by ljharb** [comment]
> If one of your libraries doesn't have any of its types @exported and you want to use some of its types, you'll be out of luck.

this is true of anything, and that’s fine. If you want access to something that’s not exported, you either ask the maintainer, fork it, or *you don’t get access*

**Comment by jespertheend** [comment]
> this is true of anything

But it doesn't have to be. A new option would be very close to just deprecating exports. And to quote sandersn:

> We're definitely not able to deprecate anything because I expect a lot of JSDoc's value is in using VS Code (or even VS!) to open ancient loose .js files and get some kind of help from jsdoc comments.

Projects with this new option enabled would lose this value just as much as if the behaviour was deprecated.

Don't get me wrong, an option is better than not fixing this issue at all. I just think there are better alternatives :)

Take JavaScripts `var` for instance: spec authors realised that `var` was not strict enough. So instead of adding yet another `"use strict";` kind of statement at the top of files, they came up with `let` and `const`. A new JSDoc tag would have the same benefits as this approach. Legacy code can still use `@typedef` while modern code uses the new tag.

**Comment by kungfooman** [comment]
> Take JavaScripts `var` for instance: spec authors realised that `var` was not strict enough. So instead of adding yet another `"use strict";` kind of statement at the top of files, they came up with `let` and `const`.

Please note that they have also come to realize that certain aspects are not viable and as a result, they are now causing syntax errors, for example:

```js
"use strict";
var static = 1;
```

![image](https://github.com/microsoft/TypeScript/assets/5236548/e2f2e0e4-d1a0-46db-a48b-935d07a3aef5)

IMO we are in the same situation here - ESM is inherently about limiting scope. If you want to make something available outside the module, you must explicitly export it using the `export` keyword.

I think we don't even need `strictJSDoc: true`, we can simply decide via `package.json` and `type === 'module'`. Because ESM implies "local by default".

**Comment by jespertheend** [comment]
> I think we don't even need `strictJSDoc: true`, we can simply decide via `package.json` and `type === 'module'`. Because ESM implies "local by default".

I'm using Deno and don't have any `package.json`.

If a new JSDoc tag is not something we are happy with (though I still think it's the best option, maybe it just needs a better name than `@localtype`), then I propose the following:

A new `@export` tag is used to mark types as exported. Files without any `@export` tags will export all types, like they have always done. But any file that contains at least one `@export` tag will only export the specified types.
This approach is similar to how TypeScript currently detects module files:

```js
   await Promise.resolve();
// ^^^^^--- 'await' expressions are only allowed at the top level of a file when that file is a module,
// but this file has no imports or exports. Consider adding an empty 'export {}' to make this file a module.
```

Doing it like this wouldn't break legacy code, allowing you to gradually update your codebase, old libraries would keep working, and you wouldn't need to configure anything.

The only downside is that this might be somewhat unexpected, but an error message similar to the 'top level await' one could take care of that.

**Comment by kungfooman** [comment]
> A new `@export` tag is used to mark types as exported. Files without any `@export` tags will export all types, like they have always done.

Then a file using this style still doesn't work: https://github.com/microsoft/TypeScript/issues/46011#issuecomment-1538383663

Or how could that be detected? Your idea is nice and I like it and we just make a special case for `export * from ...` and if any of those files export a `@typedef`, then we are following ESM guidelines?

**Comment by ljharb** [comment]
> I think we don't even need `strictJSDoc: true`, we can simply decide via `package.json` and `type === 'module'`. Because ESM implies "local by default".

this would preclude CJS packages from having the benefit, which is my entire use case.

**Comment by jespertheend** [comment]
I would say the use of `export * from ...` shouldn't have any influence of the `@typedef`s inside that file. I.e. `@typedef`s from `foo.js` would still get exported when `export * from "./foo.js"` is used in `bar.js`, whether or not `bar.js` has an `@exports` tag.

> Edit: (Now I kinda have a fear that people will just export useless dummies simply because there is no good option)

That's why I prefer a new JSDoc tag :)

**Comment by stoicbuddha** [comment]
@import doesn't re-export the types, ex:

`@import {MyController} from "../controllers/my"`

This works in VS Code for me as well, although I think it's currently only available in the Nightly build. You can also use it to import multiple typedefs from the same file, ex:

`@import {MyController, SomeOtherController} from "../controllers/my"`

**Comment by kungfooman** [comment]
@stoicbuddha I see the value of `@import`, but how does it help with this `@typedef` `export * from ...` issue?

**Comment by stoicbuddha** [comment]
@kungfooman As far as I can tell, it solves the original issue, which is that `@typedef` exports anything you import with it. `@import` can be used to avoid that problem. I could be wrong, but I *think* `export * from ...` is a different issue? I may just be misunderstanding it.

**Comment by stoicbuddha** [comment]
@ljharb Based on my understanding of the OP, the issue is that using `@typedef` to import a type, ex:

`@typedef {import("./types").SomeType} SomeType`

then re-exports the type from that file.

Using `@import` works the same way but does *not* re-export in this manner, which solves that problem. IMO, `@typedef` *should* export, as you're defining the type (even if it's via import), while `@import` should not, as it's just an import.

I'm having the exact same issue as the OP regarding a bunch of exports from within my project because of `@typedef` which is how I discovered that `@import` exists.

**Comment by ljharb** [comment]
@stoicbuddha that's true if you are importing as part of the typedef. but typedefs aren't just for imports - `@typedef {{ foo: number }} FooType` eg also would export FooType, and that's undesirable.

It's great to know that `@import` can replace that one use case of `typedef`, but there's many others :-)

**Comment by stoicbuddha** [comment]
@ljharb In that instance, wouldn't it make more sense to either:

A) Use `@type` on the value you need to type, ex: `@type {{foo: number}}`?

B) Add `@typedef {{ foo: number }} FooType` to a separate types file and use `@import` to bring it in?

**Comment by ljharb** [comment]
@stoicbuddha for A, no, because I want to reuse the defined type in multiple places, and for B no, because i don't want it accessible from outside the file, and if it's importable, then it's accessible anywhere.

**Comment by stoicbuddha** [comment]
@ljharb I'm confused as to why the accessibility is an issue. Can you clarify the use case in which you define a type that is specific to code in a given file that gets imported somewhere else? I'm guessing you have some experience with that where I don't, and I'm not trying to discount it, I'm just not clear on why that would be happening.

**Comment by ljharb** [comment]
@stoicbuddha it's a type i don't want directly usable outside the file. The use case is "i don't want the name of the type to be part of my semver-compliant public API". It's the same reason you wouldn't want every variable in a module automatically become exported :-)

**Comment by stoicbuddha** [comment]
@ljharb I understand what the effects are, but I think we'd agree that a type and a variable are wildly different levels of dangerous in this context; a variable is actually usable, whereas a type is just evaluated by the TS compiler and has no real bearing on the code itself.

What I'm trying to ascertain is: what is the real-world instance where you have a `@typedef` in a separate file, with a type specific to that file, which is imported for use elsewhere by yourself or another developer unintentionally?

**Comment by ljharb** [comment]
Given editor autocompletion, it happens a lot. Additionally, in a published package, if i rename an exported type it's a semver-major change - the level of danger there is *precisely* as dangerous as for a variable because it can fail CI pipelines.

**Comment by mhofman** [comment]
I just ran into this again today. Needed to define a local (to my module file) type to be used as an implementation detail in a set of other exported functions. Defining top level made sense to avoid duplication, but it doesn't need to be exported.

**Comment by natsuki-engr** [comment]
I'd like to bring up the issue of unwanted `@typedef` exports again. As @jespertheend noted, internal types often pollute the public API and Intellisense, which creates unnecessary noise for consumers of the package.

We’ve all seen that an `@export` approach (private by default) would be ideal for ESM, but changing the default behavior for `@typedef` is clearly not feasible due to breaking changes. We need an opt-out mechanism that doesn't break existing code.

@andrewbranch's point about TypeScript taking responsibility for its JSDoc integration remains the most relevant: since we use JSDoc for type-checking, managing visibility within the compiler is a practical necessity.

Here is a comparison of current ideas for this opt-out:
| Tag | Pros | Cons / Issues |
| :--- | :--- | :--- |
| **`@export`** | Aligns perfectly with ESM principles (explicit is better than implicit). | Requires a breaking change to make `@typedef` private by default, which is likely non-viable. |
| **`@localtype`** | Very specific to types. | Requires duplicating `@typedef` parser; needs `@localcallback` too; [hard to discover](https://github.com/microsoft/TypeScript/issues/46011#issuecomment-1781582033) |
| **`@local`** | Clear, concise, and follows the precedent of scope-related tags like `@private` or `@package`. | Not a standard JSDoc tag (yet). |

While `@import` in TS 5.5 is helpful, it doesn't solve the issue of defining local helper types that shouldn't be exported, as discussed by @stoicbuddha, @ljharb, and @mhofman.

`@local` appears to be the most pragmatic solution to allow clean API surfaces without breaking the ecosystem. Is there a path forward to formalize this?

**Comment by natsuki-engr** [comment]
Regarding the `@local` tag, a point that hasn't been discussed is how to handle cases where a type marked with `@local` becomes part of an exported type or function (such as when it is specified in a function signature).

In existing TypeScript, type aliases that are not exported are included in signatures as they are, so I believe it should be fine to follow that specification.

Below is my implementation, and I have confirmed that it can be included in signatures.
https://github.com/natsuki-engr/TypeScript/pull/1/changes#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8

**Comment by natsuki-engr** [comment]
Choosing to introduce a new tag like `@jsdoc-strict-export`, which disables automatic exports for an entire file and works in conjunction with an `@export` tag, might be a more appropriate option, since it would minimize the impact on existing codebases when automatic `@typedef` exports are eventually deprecated in the future update.

The challenge with this approach, however, is that no existing TS tag affects an entire file in this manner. While `@ts-check` may look like a file-level tag, it is technically a TS directive rather than JSDoc, so introducing tags — or TS directives — at this scope is probably something that warrants broader discussion.
